### PR TITLE
Soundness for vow(Js.Promise.t('a))  and syntax upgrade to 3

### PR DESCRIPTION
**Pull request by ncthbrt**

### CODE REVIEW

File: .bsb.lock [hunk 1]
```diff
@@ -0,0 +1 @@
+19483
```
Comment (wokalski) [inline .bsb.lock]: Should we include `.bsb.lock`

File: src/vow.re [hunk 2]
```diff
@@ -1,102 +1,100 @@
...
-    promise: Js.Promise.catch (fun _ => (handler ()).promise) vow.promise
-  };
-  let wrap promise => {promise: promise};
-  let unsafeWrap promise => {promise: promise};
-  let unwrap {promise} => promise;
+  let onError = (handler, vow) => Js.Promise.catch((_) => handler(), vow);
+  let wrap = (promise) => Js.Promise.then_((res) => Js.Promise.resolve(innerReturn(res)), promise);
+  let unsafeWrap = (promise) =>
+    Js.Promise.then_((res) => Js.Promise.resolve(innerReturn(res)), promise);
+  let unwrap = (promise) => Js.Promise.then_((x) => Js.Promise.resolve(x.inner), promise);
```
Comment (wokalski) [inline src/vow.re]: I'd simply return `x` here and have the returned type as `inner`. It'd be a breaking change though.

**Comment by wokalski** [comment]
This is clever. I like it, but I'm not sure what we should do about `unwrap`.

File: .bsb.lock (hunk 1 above)
Comment (ncthbrt) [reply .bsb.lock]: Oh sorry probably not. Think it was introduced as part of the move to reason 3

Comment (ncthbrt) [reply .bsb.lock]: I've removed it

File: src/vow.re (hunk 2 above)
Comment (ncthbrt) [reply src/vow.re]: The problem with that is that inner('a) is a hidden by the interface, so we'd have to expose the implementation in order to return `Js.Promise.t(inner('a))`
I'm happy to do so, but not sure of the advantage.

Comment (wokalski) [reply src/vow.re]: the advantage would be that we just return identity so that `unwrap` is simply noop.

**Comment by ncthbrt** [comment]
@wokalski I've updated vow to box values and promises. This should remove the other sources of unsoundness. I also made unwrap return a type of `Js.Promise.t(container('a))` where container is a record containing a single field: `value`.

**Comment by ncthbrt** [comment]
@wokalski Just pinging to inform you that I've resolved merge conflicts.

**Comment by wokalski** [comment]
@ncthbrt I got to be honest here. I attempted to review this twice but because the syntax change is mixed in, I cannot possibly review it. Would you mind submitting a PR with syntax change first?

I'm sorry for the late response but I attempted twice and it was just too much.

**Comment by ncthbrt** [comment]
Sure. Sorry @wokalski. Thanks for the feedback. I'll submit them separately.

**Comment by wokalski** [comment]
Thank you @ncthbrt. I'm usually quick to respond so I'll review it quickly!

**Pull request closed without merging**